src/common.cpp

```diff
@@ -76,16 +74,32 @@ std::vector<shape::dynamic_dimension> compute_broadcasted_dyn_dims(shape s0, sha
                        {
                            return b;
                        }
+                       else if(a.within_range(b))
+                       {
+                           return shape::dynamic_dimension{a.min, a.max};
```

Comment:
It seems strange a bit. Broadcasts are generally used to expand dimensions's range. But here for the dynamic case it picks one that is smaller in range. 
is there a good explanation on why this would be necessary ?

Reply:
I read earlier comments that it's mainly about handling {0, max_int} case. I agree with @bpickrel  that we should document this somewhere that explains why  {0, max_int} shape arises in certain cases and why broadcasting this way would be correct solution.

Reply:
docstring explains the logic. I was looking more of answer to design decision. "broadcast" is not an appropriate word because the context in which the word "broadcast" is used is different compared to logic used here. I don't have better alternative suggestion though. may be call it `compute_common_dyn_dims_for_dot` ? but then `broadcast_for_dot` would require a new name as well.

Reply:
>  of answer to design decision 

I mean it would be helpful to put some sentences in what specific case this type of broadcasting is useful and why it's necessary to do it this way.

Reply:
> broadcast" is not an appropriate word because the context in which the word "broadcast" is used is different compared to logic used here. 
It actually is broadcasting, it just so happens that matching `dynamic_dimensions` can be more lenient. At runtime the static dimensions have to be the same (unless you have a fixed 1 dimension). The only way that works is if the output dynamic_dimension range is the smaller one. There's technically an edge case with ranges that include 1, but it's pretty contrived.

Reply:
I like this explanation